JS: Handle calls to bound functions

asgerf opened the pull request:
Extends our call graph to deal with calls to bound functions, that is, functions created by a partial invocation such as `fn.bind(x, y)`. For example:

```js
function foo(x, y) {}
let fn = foo.bind(this, 0);
fn('hey'); // 'hey' flows to parameter 'y'
```

It is not enough to simply add these to `getACallee` and `FlowSteps::calls()` since that wouldn't take into account that the argument-to-parameter mapping is not 1:1.

So "bound calls" are handled as a first-class concept in the data flow libraries. We could generalize ordinary calls to be seen as "bound calls with zero bound arguments and unbound receiver", but since the overwhelming majority of calls are plain calls, it seemed wasteful.

We already type-track the result of partial invocations, for use in library models that require aggressive tracking into callbacks (`getABoundCallbackParameter`). We now use the same type-tracking to create the call graph for bound calls.

To avoid spurious flow, we distinguish cases where a bound function was tracked into a call from those that weren't. If it wasn't tracked into a call, we add a call edge that's assumed to hold in any context. If it was tracked into a call, we use our existing handling of callbacks to track the flow.

[Evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/js/returned-partial-call_1581212583919) looks fine.

esbena commented:
This seems fine, but before I check the detail, I would like to confirm that the following is intentional:

We do not use type tracking for callees in the callgraph, we only use ordinary dataflow and then type tracking for an eventual class/instance. But this change will permit type tracking for the bound function callees - right? 
Semantically, it is fine, but it is inconsistent architecturally IMO. As a consequence we will get better results for `f = bind((x) => sink(x))` than for `f = (x) => sink(x)`.

asgerf commented:
> but it is inconsistent architecturally IMO

Indeed. I've been meaning to track functions as well, but this will require a separate evaluation.

esbena commented on javascript/ql/src/semmle/javascript/dataflow/Configuration.qll:
```diff
@@ -751,7 +751,7 @@ private predicate flowThroughCall(
   exists(Function f, DataFlow::ValueNode ret |
     ret.asExpr() = f.getAReturnedExpr() and
-    calls(output, f) and // Do not consider partial calls
+    (calls(output, f) or callsBound(output, f, _)) and // Do not consider partial calls
```
We have four occurrences of the pair `calls(output, f) or callsBound(output, f, _)` in this commit, are they worth moving into a predicate?

asgerf commented on javascript/ql/src/semmle/javascript/dataflow/internal/CallGraphs.qll:
```diff
@@ -100,8 +100,12 @@ module CallGraph {
+    exists(DataFlow::TypeTracker t |
+      result = getABoundFunctionReferenceAux(function, boundArgs, t) and
+      t.end() and
+      contextDependent = t.hasCall()
```
There are two kinds of call edges: those that hold in all contexts, and those that are context-dependent. This boolean indicates which one it is. There is no reason to bind it to both `true` and `false` in one of the cases.

If you're thinking of `contextDependent` as an _input_ to the predicate somehow, I can see how your reasoning makes sense, but predicates don't really have inputs.

asgerf commented on javascript/ql/src/semmle/javascript/dataflow/internal/FlowSteps.qll:
```diff
@@ -135,6 +148,14 @@ private module CachedSteps {
+    exists(int boundArgs, int i, Parameter p |
+      callsBound(invk, f, boundArgs) and
+      f.getParameter(boundArgs + i) = p and
+      not p.isRestParameter() and
```
Hm, it might be better to change `getParameter(int)` to not return rest parameters by default. Let's discuss elsewhere.

esbena replied:
https://github.com/github/codeql-javascript-team/issues/69

esbena commented:
Implementation LGTM then.

> Do we want a generic change note for "improved callgraph"?

asgerf commented:
Rebased to fix a conflict and added change note.

I'd like to run a another short evaluation since the other PR involving partial invocations has landed.

asgerf commented:
[Evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/js/returned-partial-call_1582553792840) still looks good.

@esbena I rephrased the change note a bit, could you take a quick look?

esbena commented on change-notes/1.24/analysis-javascript.md:
```diff
@@ -13,9 +13,9 @@
-* Calls can now be resolved to class members in more cases, leading to more results from the security queries.
-
-* Calls through partial invocations such as `.bind()` are now analyzed more precisely, leading to more results from the security queries.
+* The call graph construction has been improved a few ways, leading to more results from the security queries:
```
Totally optional suggestion, but something to consider next time:

```suggestion
* The call graph construction has been improved, leading to more results from the security queries:
```
